src/writer.rs

```diff
@@ -745,123 +745,166 @@ impl<W: io::Write> Writer<W> {
+    /// | struct | `struct Foo { a: u8, b: bool }` | `Foo { a: 5, b: true }` | `5,true` |
     ///
-    /// Simple enums in Rust can be serialized. Namely, enums must either be
-    /// variants with no arguments or variants with a single argument. For
-    /// example, to serialize a field from either an integer or a float type,
-    /// one can do this:
+    /// If `has_headers` is `false`, then there are no additional restrictions;
+    /// types can be nested arbitrarily. For example:
     ///
     /// ```
     /// extern crate csv;
     /// #[macro_use]
     /// extern crate serde_derive;
     ///
     /// use std::error::Error;
-    /// use csv::Writer;
+    /// use csv::WriterBuilder;
     ///
     /// #[derive(Serialize)]
     /// struct Row {
     ///     label: String,
-    ///     value: Value,
-    /// }
-    ///
-    /// #[derive(Serialize)]
-    /// enum Value {
-    ///     Integer(i64),
-    ///     Float(f64),
+    ///     values: Vec<f64>,
     /// }
     ///
     /// # fn main() { example().unwrap(); }
     /// fn example() -> Result<(), Box<Error>> {
-    ///     let mut wtr = Writer::from_writer(vec![]);
+    ///     let mut wtr = WriterBuilder::new()
+    ///         .has_headers(false)
+    ///         .from_writer(vec![]);
     ///     wtr.serialize(Row {
     ///         label: "foo".to_string(),
-    ///         value: Value::Integer(3),
-    ///     })?;
-    ///     wtr.serialize(Row {
-    ///         label: "bar".to_string(),
-    ///         value: Value::Float(3.14),
+    ///         values: vec![1.1234, 2.5678, 3.14],
     ///     })?;
     ///
     ///     let data = String::from_utf8(wtr.into_inner()?)?;
     ///     assert_eq!(data, "\
-    ///label,value
-    ///foo,3
-    ///bar,3.14
-    ///");
+    /// foo,1.1234,2.5678,3.14
+    /// ");
     ///     Ok(())
     /// }
     /// ```
-    pub fn serialize<S: Serialize>(&mut self, mut record: S) -> Result<()> {
-        match self.state.header {
-            HeaderState::None | HeaderState::DidNotWrite => {
-                serialize(self, record, false, false)?;
-                self.write_terminator()?;
-            }
-            HeaderState::DidWrite => {
-                serialize(self, record, false, true)?;
-                self.write_terminator()?;
-            }
-            HeaderState::Write => {
-                let did = serialize(self, &mut record, true, false)?;
-                self.state.header =
-                    if did {
-                        HeaderState::DidWrite
-                    } else {
-                        HeaderState::DidNotWrite
-                    };
+    ///
+    /// However, if `has_headers` were enabled in the above example, then
+    /// serialization would return an error. Speficially, when `has_headers` is
+    /// `true`, there are two restrictions:
+    ///
+    /// 1. Named field values in structs must be scalars.
+    ///
+    /// 2. All scalars must be named field values in structs.
```

Comment:
Does this get formatted as a list? You might need to delete the empty line before (2).